File: hexameter.core/src/main/java/org/codetome/hexameter/core/internal/impl/layoutstrategy/TrapezoidGridLayoutStrategy.java
package org.codetome.hexameter.core.internal.impl.layoutstrategy;

import org.codetome.hexameter.core.api.CubeCoordinate;
import org.codetome.hexameter.core.api.HexagonalGridBuilder;

import java.util.ArrayList;
import java.util.List;

public final class TrapezoidGridLayoutStrategy extends GridLayoutStrategy {

    @Override
    public Iterable<CubeCoordinate> fetchGridCoordinates(final HexagonalGridBuilder builder) {
        final List<CubeCoordinate> coords = new ArrayList<>();
        for (int gridZ = 0; gridZ < builder.getGridHeight(); gridZ++) {
            for (int gridX = 0; gridX < builder.getGridWidth(); gridX++) {
                coords.add(CubeCoordinate.fromCoordinates(gridX, gridZ));
            }
        }
        return coords;
    }

    @Override
    public boolean checkParameters(final int gridHeight, final int gridWidth) {
        return checkCommonCase(gridHeight, gridWidth);
    }
}

File: mixite2.core/src/commonMain/kotlin/org/hexworks/mixite2/core/internal/impl/layoutstrategy/TrapezoidGridLayoutStrategy.kt
package org.hexworks.mixite2.core.internal.impl.layoutstrategy

import org.hexworks.mixite2.core.api.CubeCoordinate
import org.hexworks.mixite2.core.api.HexagonalGridBuilder
import org.hexworks.mixite2.core.api.contract.SatelliteData

class TrapezoidGridLayoutStrategy : GridLayoutStrategy() {

    override fun fetchGridCoordinates(builder: HexagonalGridBuilder<out SatelliteData>): Iterable<org.hexworks.mixite2.core.api.CubeCoordinate> {
        val coords = ArrayList<org.hexworks.mixite2.core.api.CubeCoordinate>(builder.getGridHeight() * builder.getGridWidth())
        for (gridZ in 0 until builder.getGridHeight()) {
            for (gridX in 0 until builder.getGridWidth()) {
                coords.add(org.hexworks.mixite2.core.api.CubeCoordinate.fromCoordinates(gridX, gridZ))
            }
        }
        return coords
    }

    override fun checkParameters(gridHeight: Int, gridWidth: Int): Boolean {
        return checkCommonCase(gridHeight, gridWidth)
    }

    override fun getName(): String {
        return "TRAPEZOID"
    }
}
